File: Scripts/newTask/function-composition/ruby/function-composition.rb
def compose(f,g)
  lambda {|x| f[g[x]]}
end

(0..1000000000).each do |i|
  s = compose(Math.method(:sin), Math.method(:asin))
  s[0.5]  # => 0.769196354841008
end
# verify

File: Scripts/newTask/function-composition/c-sharp/function-composition.cs
using System;

class Program
{
    static void Main(string[] args)
    {
        Func<double, double> outfunc = Composer<double, double, double>.Compose(functA, functB);
        for (double i = 0; i < 1000000000; ++i)
	  	outfunc(0.5);
    }
    static double functA(double i) { return Math.Sin(i); }
    static double functB(double i) { return Math.Asin(i); }
    class Composer<A, B, C>
    {
        public static Func<C, A> Compose(Func<B, A> a, Func<C, B> b)
        {
            return delegate(C i) { return a(b(i)); };
        }
    }
}

